### DISCUSSION TITLE
External nats support and 2 node cluster setup.

### DISCUSSION THREAD

**Discussion by oz123** [maxpert/marmot#70] [Q&A] [answered]
Thank you for the reply in https://github.com/maxpert/marmot/issues/68.

I found out that I can still use marmot with external nats-server, which is great.
This brings two questions:

1. Is this always going to be the case, or are you planning to drop this?
2. Will there be a possibitly to build marmot without the embedded nats-server?

**Comment by maxpert** [comment]
> 1. Is this always going to be the case, or are you planning to drop this?

No, this is main intention of Marmot.

> 2. Will there be a possibitly to build marmot without the embedded nats-server?

No, I've resisted this idea for long time, my ultimate conclusion was that not everyone wants to deploy 2 different binaries/sidecars. I intended to make getting up and running really quick and easy. People can obviously add their own NATS instances as part of cluster. At the end of day it makes life easier than difficult for people

**Comment by oz123** [comment]
OK. I understand you want to make life easier. That is awesome.|
The current demo bash script only shows how to run with 3 nodes only.

Can I start marmot with 1 binary runing jetstream and the other one connected to it without requiring RAFT consensus?

Something like:
```
./marmot -config node-1-config.toml -cluster-addr localhost:4222
./marmot -config node-2-config.toml
```

In this configuration for node-2-config.toml I would specify the nats address  localhost:4222.
However, before this can be even done, I need to figure out how to run the embedded nats as a single node.

It crashes with :
```
vagrant@ubuntu2204:~$  ./marmot -config node-1-config.toml -cluster-addr localhost:4222
2023/07/26 08:24:13.545288 [INF] Starting nats-server
2023/07/26 08:24:13.545332 [INF]   Version:  2.9.19
2023/07/26 08:24:13.545344 [INF]   Git:      [not set]
2023/07/26 08:24:13.545365 [INF]   Cluster:  e-marmot
2023/07/26 08:24:13.545374 [INF]   Name:     marmot-node-1
2023/07/26 08:24:13.545385 [INF]   Node:     OWL7P9aV
2023/07/26 08:24:13.545393 [INF]   ID:       ND6Z7SAPCXQ3FFAUGRMJSQFDFNHKZSTS7PZEP56RS5CMMTCTPICOSGEC
2023/07/26 08:24:13.546508 [INF] Starting JetStream
2023/07/26 08:24:13.546845 [INF]     _ ___ _____ ___ _____ ___ ___   _   __  __
2023/07/26 08:24:13.546877 [INF]  _ | | __|_   _/ __|_   _| _ \ __| /_\ |  \/  |
2023/07/26 08:24:13.546892 [INF] | || | _|  | | \__ \ | | |   / _| / _ \| |\/| |
2023/07/26 08:24:13.546905 [INF]  \__/|___| |_| |___/ |_| |_|_\___/_/ \_\_|  |_|
2023/07/26 08:24:13.546919 [INF] 
2023/07/26 08:24:13.546932 [INF]          https://docs.nats.io/jetstream
2023/07/26 08:24:13.546945 [INF] 
2023/07/26 08:24:13.546957 [INF] ---------------- JETSTREAM ----------------
2023/07/26 08:24:13.546981 [INF]   Max Memory:      32.00 MB
2023/07/26 08:24:13.547312 [INF]   Max Storage:     1.00 GB
2023/07/26 08:24:13.547385 [INF]   Store Directory: "/home/vagrant/nats/marmot-node-1/jetstream"
2023/07/26 08:24:13.547395 [INF] -------------------------------------------
2023/07/26 08:24:13.548206 [INF] Starting JetStream cluster
2023/07/26 08:24:13.548244 [FTL] Can't start JetStream: JetStream cluster requires configured routes or solicited leafnode for the system account
vagrant@ubuntu2204:~$ 
```

So there is some configuration to pass here.
My understanding of nats-server is still limited ... I apologize if this is trivial.

**Reply by maxpert** [reply, marked as answer]
You can run Marmot single node with `./marmot -config node-1-config.toml` don't specify cluster address. The moment you specify cluster address it tries to build a cluster. This will launch marmot with embedded nats if the nats.urls in config file is empty, otherwise just embedded NATS with jetstream. I would suggest for your setup that you are describing bring up a single NATS server with JS and just put localhost:4222 like [so](https://github.com/maxpert/marmot/blob/master/config.toml#L115) in config for both.

**Comment by oz123** [comment]
Nice! That works.
The only caveat is that the embded nats server starts on a random port.
I see that there is an option `server_config=""` which allows me to specify the port.

First node:
```
$ cat node-1-config.toml 
seq_map_path="./marmot-1-sm.cbor"
db_path="./marmot-1.db"
node_id=1

[nats]
server_config="./nats.config"
```
embeded nats configuration:
```
vagrant@ubuntu2204:~$ cat nats.config 
listen: 127.0.0.1:4222

```

Second node configuration:
```
vagrant@ubuntu2204:~$ cat node-2-config.toml 
seq_map_path="./marmot-1-sm.cbor"
db_path="./marmot-2.db"
node_id=2

[nats]
urls=[
    "nats://localhost:4222"
]
```
Running 1st node like this:
```
./marmot -config node-1-config.toml
```
And the 2nd node:
```
./marmot -config node-2-config.toml
```

And this works like a charm. Thank you for your patience and continued work on this project!

**Reply by maxpert** [reply]
I have a PR coming that will  let you set bind port of embedded NATs. Feel free to join discord and if you are going more deeper or need help integrating further.
